Homework 4

antoska53 opened the pull request.

antoska53 committed:
add RecyclerView

добавил ресайкл вью, пока что ничего не работает.

antoska53 committed:
add RcyclerView and ClickListener

добавлен ресайклервью, теперь он работает. Добавлен клик листенер, открывающий детальное описание фильма

antoska53 committed:
add RecyclerView for Actor

добавлен ресайклер вью для актеров (fragment_movie_details)

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/ActorAdapter.kt:
```diff
@@ -0,0 +1,22 @@
+import androidx.recyclerview.widget.RecyclerView
+
+class ActorAdapter: RecyclerView.Adapter<ActorViewHolder>() {
+    private var listActor = listOf("Robert Downey Jr", "Mark Ruffalo", "Chris Evans", "Chris Hemsworth", "Robert Downey Jr", "Mark Ruffalo", "Chris Evans", "Chris Hemsworth")
```
слишком длинное, лучше в столбец, тк читать не удобно

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/ActorAdapter.kt:
```diff
@@ -0,0 +1,22 @@
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ActorViewHolder {
+        var view = LayoutInflater.from(parent.context).inflate(R.layout.view_holder_actor, parent, false)
+        return ActorViewHolder(view)
```
можно в 1 строку

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/ActorAdapter.kt:
```diff
@@ -0,0 +1,22 @@
+
+    override fun getItemCount(): Int {
+        return listActor.size
+    }
```
можно в 1 строку

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/ActorViewHolder.kt:
```diff
@@ -0,0 +1,32 @@
+
+    fun onBind(nameActor: String) {
+        when (nameActor) {
+            "Robert Downey Jr" -> {
+                ivActor.setImageResource(R.drawable.robert_downey)
+                tvActorName.setText(R.string.robert_downey_jr)
+            }
+            "Mark Ruffalo" -> {
+                ivActor.setImageResource(R.drawable.mark_ruffalo)
+                tvActorName.setText(R.string.mark_ruffalo)
+            }
+            "Chris Evans" -> {
+                ivActor.setImageResource(R.drawable.chris_evans)
+                tvActorName.setText(R.string.chris_evans)
+            }
+            "Chris Hemsworth" -> {
+                ivActor.setImageResource(R.drawable.chris_hemsworth)
+                tvActorName.setText(R.string.chris_hemsworth)
```
оочень странный подход, так не делают, текст в любой момент может измениться
в onBind обычно приходит какой-то объект, из которого потом заполняются поля, погуглите примеры работы с RecyclerView

antoska53 replied:
@jokerxt это понятно, но у нас же в этой домашке жестко прописаны имена и картинки, мне просто нечего передавать в онбаинд, если только специально не оборачивать имя, но зачем...

jokerxt replied:
а затем, что потом надо будет) надо думать наперед) да и такого как вы написали не может быть никогда

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/FragmentMoviesDetails.kt:
```diff
@@ -18,13 +23,13 @@ private const val ARG_PARAM2 = "param2"
 class FragmentMoviesDetails : Fragment() {
     // TODO: Rename and change types of parameters
-    private var param1: String? = null
+    private var nameMovie: String? = null
```
используйте by lazy

antoska53 replied:
@jokerxt не совсем понимаю как его здесь использовать, ведь мне надо пулучить данные из бандла в onCreate.

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/FragmentMoviesDetails.kt:
```diff
@@ -18,13 +23,13 @@ private const val ARG_PARAM2 = "param2"
         arguments?.let {
-            param1 = it.getString(ARG_PARAM1)
+            nameMovie = it.getString(NAME_MOVIE)
             param2 = it.getString(ARG_PARAM2)
```
все можно удалить, если использовать by lazy

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/FragmentMoviesDetails.kt:
```diff
@@ -36,7 +41,38 @@ class FragmentMoviesDetails : Fragment() {
+        var tvMovieDescription: TextView = view.findViewById(R.id.description)
+
+        when (nameMovie) {
+            "avengers" -> {
+                ivImageMovie.setImageResource(R.drawable.avengers)
+                tvMovieDescription.setText(R.string.description_avengers)
+            }
+            "tenet" -> {
+                ivImageMovie.setImageResource(R.drawable.tenet_movie)
+                tvMovieDescription.setText(R.string.description_tenet)
+            }
+            "black widow" -> {
+                ivImageMovie.setImageResource(R.drawable.black_widow_movie)
+                tvMovieDescription.setText(R.string.description_black_widow)
+            }
+            "wonder woman" -> {
+                ivImageMovie.setImageResource(R.drawable.wonder_woman_movie)
+                tvMovieDescription.setText(R.string.description_wonder_woman)
+            }
```
опять странно

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/FragmentMoviesDetails.kt:
```diff
@@ -36,7 +41,38 @@ class FragmentMoviesDetails : Fragment() {
     ): View? {
         // Inflate the layout for this fragment
-        return inflater.inflate(R.layout.fragment_movie_details, container, false)
+        var view: View = inflater.inflate(R.layout.fragment_movie_details, container, false)
```
фрагмент можно вообще создать используя конструктор с параметром куда передать layoutResId, погуглите

antoska53 replied:
@jokerxt читал что нельзя создавать фрагменты через конструктор с параметрами, так как андроид ничо не знает о мои параметрах, и при смене конфигурации всё может полететь к чертям =)

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/FragmentMoviesDetails.kt:
```diff
@@ -53,7 +89,7 @@ class FragmentMoviesDetails : Fragment() {
         fun newInstance(param1: String, param2: String) =
             FragmentMoviesDetails().apply {
                 arguments = Bundle().apply {
```
можно использовать bundleOf

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/FragmentMoviesList.kt:
```diff
@@ -43,15 +45,25 @@ class FragmentMoviesList : Fragment() {
+        var recycler: RecyclerView = view.findViewById(R.id.recycler_movie)
+        recycler.layoutManager = GridLayoutManager(view.context,2)
+        recycler?.adapter = MovieAdapter(listener)
+        var dividerItemDecoration :MovieItemDecoration = MovieItemDecoration(20)
```
выравнивание

jokerxt replied:
и магическая цифра 20, что это откуда, выносите в ресурсы или в константы

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/FragmentMoviesList.kt:
```diff
@@ -43,15 +45,25 @@ class FragmentMoviesList : Fragment() {
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        var recycler: RecyclerView = view.findViewById(R.id.recycler_movie)
+        recycler.layoutManager = GridLayoutManager(view.context,2)
```
выравнивание
и непонятная цифра 2

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/FragmentMoviesList.kt:
```diff
@@ -43,15 +45,25 @@ class FragmentMoviesList : Fragment() {
-        imageView = view?.findViewById(R.id.mask_avengers_movie)
-        imageView?.setOnClickListener(View.OnClickListener { listener?.itemClicked(FragmentMoviesDetails.newInstance("", "")) })
+        //imageView = view?.findViewById(R.id.mask_avengers_movie)
+       //imageView?.setOnClickListener(View.OnClickListener { listener?.itemClicked(FragmentMoviesDetails.newInstance("", "")) })
```
зачем заливать комментарии

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/MainMovieActivity.kt:
```diff
@@ -12,7 +12,7 @@ class MainMovieActivity : AppCompatActivity(), FragmentMoviesList.Companion.List
 
         if (savedInstanceState == null){
-            changeFragment(FragmentMoviesList.newInstance("",""))
+            changeFragmentNull(FragmentMoviesList.newInstance("",""))
```
почему Null?

antoska53 replied:
@jokerxt в нем нет добавления в бэкстек.

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/MovieAdapter.kt:
```diff
@@ -0,0 +1,28 @@
+
+    override fun getItemCount(): Int {
+        return listMovie.size
+    }
```
1 строка  override `fun getItemCount(): Int = listMovie.size`

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/MovieAdapter.kt:
```diff
@@ -0,0 +1,28 @@
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
+        var view = LayoutInflater.from(parent.context).inflate(R.layout.view_holder_movie, parent,false)
+        return MovieViewHolder(view)
```
1 строка

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/MovieAdapter.kt:
```diff
@@ -0,0 +1,28 @@
+import androidx.fragment.app.Fragment
+import androidx.recyclerview.widget.RecyclerView
+
+class MovieAdapter(private val listener: FragmentMoviesList.Companion.Listener?): RecyclerView.Adapter<MovieViewHolder>() {
```
лучше не использовать Listener из компаньона, сделайте нормальный иначе может что-нибудь утечь

antoska53 replied:
@jokerxt нормальный это какой?

antoska53 replied:
@jokerxt видел такие примеры в книге по java для андроида, и в лекциях на ютубе. В котлине надо как то по другому делать? Не понятно, если не делать его копаньеном, то тогда делать просто отдельный интерфейс? Но зачем он нужен если он используется только внутри этого фрагмента и больше нигде?

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/MovieItemDecoration.kt:
```diff
@@ -0,0 +1,33 @@
+
+    constructor(offset: Int) : super() {
+        this.offset = offset
+    }
```
нужен всего 1 конструктор, можно его вынести наверх и не нужен никакой var offset
`class MovieItemDecoration(private val offset: Int)`

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/MovieViewHolder.kt:
```diff
@@ -0,0 +1,37 @@
+    private val tvNameMovie: TextView = itemView.findViewById(R.id.tv_name)
+
+    fun onBind(nameMovie: String, listener: FragmentMoviesList.Companion.Listener?) {
+        itemView.setOnClickListener(View.OnClickListener {
```
можно без View.OnClickListener

jokerxt commented on app/src/main/java/ru/myacademyhomework/myfilms/MovieViewHolder.kt:
```diff
@@ -0,0 +1,37 @@
+        })
+        imageMovie.setImageResource(R.drawable.avengers_movie)
+        when (nameMovie) {
+            "avangers" -> {
+                imageMovie.setImageResource(R.drawable.avengers_movie)
+                tvNameMovie.setText(R.string.movie_name_avengers)
+            }
+            "tenet" -> {
+                imageMovie.setImageResource(R.drawable.tenet_movie)
+                tvNameMovie.setText(R.string.movie_name_tenet)
+            }
+            "black widow" -> {
+                imageMovie.setImageResource(R.drawable.black_widow_movie)
+                tvNameMovie.setText(R.string.movie_name_black_widow)
+            }
+            "wonder woman" -> {
+                imageMovie.setImageResource(R.drawable.wonder_woman_movie)
+                tvNameMovie.setText(R.string.movie_wonder_woman)
+            }
```
опять тексты

jokerxt commented on app/src/main/res/layout/fragment_movies_list.xml:
```diff
@@ -4,6 +4,7 @@
     android:layout_height="match_parent"
     xmlns:app="http://schemas.android.com/apk/res-auto">
 
+
```
зачем?

antoska53 replied:
@jokerxt что именно зачем???

jokerxt commented on app/src/main/res/layout/view_holder_movie.xml:
```diff
@@ -0,0 +1,174 @@
+        android:textSize="8sp"
+        android:lineSpacingExtra="4sp"
+        android:text="@string/review"
+        android:textColor="#6D6D80"
```
цвет в ресурсы

jokerxt commented on app/src/main/res/layout/view_holder_movie.xml:
```diff
@@ -0,0 +1,174 @@
+        android:layout_height="wrap_content"
+        android:layout_centerHorizontal="true"
+        android:layout_centerVertical="true"
+        android:layout_marginStart="8.69dp"
```
странный отступ

jokerxt commented on app/src/main/res/layout/view_holder_movie.xml:
```diff
@@ -0,0 +1,174 @@
+        android:layout_marginBottom="4dp"
+        android:text="@string/movie_genre"
+        android:textSize="8sp"
+        android:textColor="#FF3466"
```
цвет в ресурсы

jokerxt commented on app/src/main/res/layout/view_holder_movie.xml:
```diff
@@ -0,0 +1,174 @@
+        android:layout_marginStart="8dp"
+        android:layout_marginTop="8dp"
+        android:textSize="14sp"
+        android:textColor="#ECECEC"
```
цвет в ресурсы

jokerxt commented on app/src/main/res/layout/view_holder_movie.xml:
```diff
@@ -0,0 +1,174 @@
+        android:layout_height="wrap_content"
+        android:text="@string/avengers_minutes"
+        android:textSize="8sp"
+        android:textColor="#565665"
```
цвет в ресурсы

jokerxt commented on app/src/main/res/layout/view_holder_movie.xml:
```diff
@@ -0,0 +1,174 @@
+        android:layout_height="wrap_content"
+        android:text="@string/rectangle_text"
+        android:textSize="12sp"
+        android:textColor="#FFFFFF"
```
цвет в ресурсы

antoska53 committed:
исправления по замечаниям ментора

antoska53 committed:
Исправления по замечаниям ментора

Исправлены все замечания. 
Добавлена Ленивая инициализация во фрагменты. 
Фрагменты создаются через конструктор с параметром. 
Убран listener из companion object фрагмента. Теперь это просто отдельный интерфейс
Плюс мелкие исправления

antoska53 committed:
Merge branch 'Homeworl_4_for_review' into Homework_4
